protocol/v2/qbft/controller/decided.go

```diff
@@ -45,31 +45,34 @@ func (c *Controller) UponDecided(msg *specqbft.SignedMessage) (*specqbft.SignedM
 	} else { // decide previously, add if has more signers
 		signers, _ := inst.State.CommitContainer.LongestUniqueSignersForRoundAndValue(msg.Message.Round, msg.Message.Data)
 		if len(msg.Signers) > len(signers) {
+			//TODO: we dont broadcast late decided anymore, so we should check if we can remove this
 			inst.State.CommitContainer.AddMsg(msg)
+			// TODO: update storage
 		}
 	}
 
+	if msg.Message.Height != c.Height {
```

Comment:
It's sufficient to check previous decided here, no need to check height..
We can also move this block to where it was, below future check as we want to be aligned with spec.

The only difference here should be `SaveInstance` as we want to support fullnode